man/man1/statclm.1
.TH statclm 1  "USER COMMANDS"
.SH NAME
statclm \- print statistics of clm files
.SH SYNOPSIS
.B statclm
[-metafile] [-scale s] [-version] [-verbose] [-csv] [-type {byte|short|int|float|double}] \fIfile.clm ...\fP

.SH DESCRIPTION
Program prints minimum, maximum and average of all data in a clm file.
.SH OPTIONS
.TP
-metafile
File is a JSON metafile describing the structure of the binary file.
.TP
-scale s
Values are scaled by a factor s if version of clm file is less than 2.
.TP
-version v
Set version of clm file to v. If not specified version is read from clm file.
.TP
-type {byte|short|int|float|double}
Set datatype of clm file if version of clm file is less than 3. Default is short.
.TP
-verbose
Print statistics for each year.
.TP
-csv
Print statistics in a comma-separated format.
.TP
.I file.clm
Filename(s) of clm files.
.SH EXAMPLE
.TP
Print statistics for each year:
.B statclm
-verbose temp.clm
.PP
.SH EXIT STATUS
.B statclm
returns a zero exit status if statistics of clm file could be printed.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
printclm(1), setclm(1), headersize(1), lpjml(1), cru2clm(1), grid2clm(1), cft2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de)

### Added

- GIT repository and GIT hash printed in the LPJmL banner and in output of NetCDF and JSON files.
- Options `-repo` and `-hash` added to `lpjml` to print GIT repository and hash.
- The JSON file created by `createconfig` contains now name and hash of GIT repository.
- Command `lpjml -v` prints GIT repository and hash.
- Target `hash` added to Makefile to update hash after committing changes to the GIT repository.

### Changed

- `LPJ_VERSION` macro in `lpj.h` replaced by function `getversion()`.

### Fixed

- Missing dependency on `gebuild.c` for target `lpjml` added in `src/Makefile`.
- `xiar` replaced by `ar` in `Makefile.hpc2024` in order to compile with `intel/oneAPI/2025.0.0`.
- Typos in error messages in `filesexist.c` and `cdf2soil.c`fixed.


## [5.9.15] - 2024-11-07

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- Option `-scale` added to `bin2cdf` in order to scale output.
- Option `-notime` added to `bin2cdf` in order to omit time axis.
- `time_bnds`, `lat_bnds`, and `lon_bnds` arrays added to NetCDF output for lpjml runs and the `bin2cdf` utility.
- Check for identical output filenames added in `fscanoutput.c`.

### Changed

- If `"with_days"` is set to true in the lpjml configuration file, time axis of yearly output is also set in units of days instead of years.
- Time in NetCDF output is always set in the middle between the lower and upper time boundaries.

### Removed

- Unused file `cflux_sum.c` removed.
- Conversion of obsolete file `manage.par` removed from utility `manage2js`.

### Fixed

- Separate output for each year by setting `"name" : "filename_%d"` is now written correctly (issue #360).
- Number of bands set to 1 in metafile of grid NetCDF output.
- Name of longitude/latitude variables correctly derived in `getlatlon_netcdf.c`.
- Misspelled option `-netdcdf4` corrected to `-netcdf4` in `clm2cdf.c`.
- `lpj_climber4.c` updated to compile without errors.
- Writing different variables into one NetCDF file fixed in `create_netcdf.c`. Only NetCDF outputs with one band and same time step can be written to one file.
- Formatting of man pages corrected.
- Missing `shift 1` for `-crumonthly` case added in `regridlpj`.


## [5.9.14] - 2024-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Fixed

- `pft->vscal` not set to `NDAYYEAR` in `update_annual.c`. (issue #364).
- `freezefrac2soil.c` corrected to avoid division by zero.
- size of option array for `sowing_date_option` and `crop_phu_option` now correctly specified in `fscanconfig.c`.


## [5.9.13] - 2024-09-26

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `USE_NETCDF4` compile option replaced by `"netcdf4"` boolean flag in the LPJmL configuration file. If set compression of NetCDF4 files can be enabled and PFT names are written as strings instead of character arrays.
- Formatting of man pages harmonized.
- If lpjml is compiled without `-DUSE_NETCDF` flag then `fscanconfig.c` and `fscanoutput.c` return with an error for NetCDF files specified for input/output.

### Added

- Option `-netcdf4` added to `bin2cdf`, `clm2cdf`, `country2cdf`, and `drainage2cdf` utility to enable NetCDF4 format.
- Boolean flag `"netcdf4"` added to LPJmL configuration file in order to enable NetCDF4 format.
- Option `-compress` added to `drainage2cdf` utility to enable file compression.
- Man page for `drainage2cdf` added.
- Check for valid compression value added for lpjml and utilities.
- Missing option `"global_netcdf"` added to LPJmL config file.

### Removed

- Obsolete man page for `writeregioncode()` removed.

### Fixed

- Missing deallocation of memory added in `cpl_init.c`, `fscanlandcovermap.c`, `fscanoutput.c`, `writearea.c`, `create_pft_netcdf.c`, `newgrid.c`, `fscanagtreemap.c`, `celldata.c` in case of error.
- Syntax error fixed in `update_daily.c` in IMAGE coupling.
- Typo in `README` corrected.
- `establishmentpft.c`, `cdf2coord.c` and `cdf2grid.c` modified to compile without warnings using gcc.
- Debug flag in `send_token_coupler.c` and `openoutput_coupler.c` corrected to `DEBUG_COUPLER`.
- Code changed to compile without errors/warnings for `-DIMAGE -DCOUPLED` setting.
- Calculation of offsets and counts corrected in `readintdata_netdf.c` and missing loop over bands added.


## [5.9.12] - 2024-09-24

### Contributors

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Werner von Bloh (bloh@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- added SLURM option `--cpus-per-task=1` to `lpjsubmit` template `bin/lpjsubmit_hpc` to avoid that mpirun uses several CPUs per process


## [5.9.11] - 2024-09-23

### Contributors

- author: Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

src/utils/statclm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                       s  t  a  t  c  l  m  .  c                                \n**/
/**                                                                                \n**/
/**     Program prints minimum, maximum and average value of clm files             \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-metafile] [-version v] [-scale s] [-verbose] [-csv] [-type {byte|short|int|float|double}] filename ...\n"

int main(int argc,char **argv)
{
  Header header;
  String id;
  FILE *file;
  long long size;
  size_t offset;
  Bool swap,verbose,iscsv,iserr,ismeta;
  int version,i,j,cell,iarg,year,index,setversion;
  float fmin,fmax,favg,avg,cavg,fmin2,fmax2,scale;
  float *vec;
  char *endptr,*unit=NULL;
  Type datatype;
  scale=1;
  verbose=iscsv=ismeta=FALSE;
  datatype=LPJ_SHORT;
  setversion=READ_VERSION;
  for(iarg=1;iarg<argc;iarg++)
  {
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-scale"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Argument missing after '-scale' option.\n"
                USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        scale=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-scale'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-version"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Argument missing after '-version' option.\n"
                USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        setversion=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-version'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(setversion>CLM_MAX_VERSION)
        {
          fprintf(stderr,"Version %d greater than maximum version %d supported.\n",
                  setversion,CLM_MAX_VERSION);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-type"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Argument missing after option '-type'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        index=findstr(argv[++iarg],typenames,5);
        if(index==NOT_FOUND)
        {
          fprintf(stderr,"Invalid argument '%s' for option '-type'.\n"
                  USAGE,argv[iarg],argv[0]);
          return EXIT_FAILURE;
        }
        datatype=(Type)index;
      }
      else if(!strcmp(argv[iarg],"-metafile"))
        ismeta=TRUE;
      else if(!strcmp(argv[iarg],"-verbose"))
        verbose=TRUE;
      else if(!strcmp(argv[iarg],"-csv"))
        iscsv=TRUE;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n",argv[iarg]);
        fprintf(stderr,USAGE,argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
     break;
  }
  if(argc<iarg+1)
  {
    fprintf(stderr,"Missing argument.\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  if(iscsv)
  {
    if(verbose)
    {
      if(unit!=NULL && strlen(unit))
        printf("filename,year,min (%s) ,max (%s) ,avg (%s) \n",unit,unit,unit);
      else
        printf("filename,year,min,max,avg\n");
    }
    else
    {
      if(unit!=NULL && strlen(unit))
        printf("filename,min (%s) ,max (%s) ,avg (%s) \n",unit,unit,unit);
      else
        printf("filename,min,max,avg\n");
    }
  }
  for(i=iarg;i<argc;i++)
  {
    if(ismeta)
    {
      header.scalar=1;
      header.cellsize_lon=header.cellsize_lat=0.5;
      header.firstyear=1901;
      header.firstcell=0;
      header.nyear=1;
      header.nbands=1;
      header.nstep=1;
      header.timestep=1;
      header.datatype=datatype;
      header.order=CELLYEAR;
      header.scalar=scale;
      file=openmetafile(&header,NULL,NULL,NULL,NULL,NULL,NULL,NULL,&unit,NULL,NULL,NULL,NULL,NULL,&swap,&offset,argv[i],TRUE);
      if(file==NULL)
        continue;
      if(fseek(file,offset,SEEK_CUR))
      {
        fprintf(stderr,"Error seeking in '%s' to offset %lu.\n",argv[i],offset);
        fclose(file);
        continue;
      }
    }
    else
    {
      version=setversion;
      file=fopen(argv[i],"rb");
      if(file==NULL)
      {
        fprintf(stderr,"Error opening '%s': %s\n",argv[i],strerror(errno));
        return EXIT_FAILURE;
      }
      if(freadanyheader(file,&header,&swap,id,&version,TRUE))
      {
        fclose(file);
        continue;
      }
      if(version<2)
        header.scalar=scale;
      if(version<3)
        header.datatype=datatype;
      size=getfilesizep(file)-headersize(id,version);
      if(size!=typesizes[header.datatype]*header.ncell*header.nbands*header.nyear*header.nstep)
        fprintf(stderr,"Warning: File length of '%s' does not match header, differs by %lld bytes.\n",
                argv[i],llabs(size-(long long)typesizes[header.datatype]*header.ncell*header.nbands*header.nstep*header.nyear));
    }
    fmin=HUGE_VAL;
    fmax=-HUGE_VAL;
    favg=0;
    vec=newvec(float,header.nstep*header.nbands);
    check(vec);
    iserr=FALSE;
    for(year=0;year<header.nyear;year++)
    {
      cavg=0;
      fmin2=HUGE_VAL;
      fmax2=-HUGE_VAL;
      for(cell=0;cell<header.ncell;cell++)
      {
        avg=0;
        if(readfloatvec(file,vec,header.scalar,header.nstep*header.nbands,swap,header.datatype))
        {
          fprintf(stderr,"Unexpected end of file in '%s'.\n",argv[i]);
          iserr=TRUE;
          break;
        }
        for(j=0;j<header.nstep*header.nbands;j++)
        {
          fmin=min(fmin,vec[j]);
          fmax=max(fmax,vec[j]);
          fmin2=min(fmin2,vec[j]);
          fmax2=max(fmax2,vec[j]);
          avg+=vec[j];
        }
        cavg+=avg/header.nstep/header.nbands;
      }
      if(iserr)
        break;
      favg+=cavg/header.ncell;
      if(verbose)
      {
        if(iscsv)
          printf("%s,%d,%g,%g,%g\n",argv[i],year+header.firstyear,fmin2,fmax2,cavg/header.ncell);
        else
        {
          if(argc-iarg>1)
            printf("%s: ",argv[i]);
          if(unit!=NULL && strlen(unit))
            printf("year=%d, min=%g %s, max=%g %s, avg=%g %s\n",year+header.firstyear,fmin2,unit,fmax2,unit,cavg/header.ncell,unit);
          else
            printf("year=%d, min=%g, max=%g, avg=%g\n",year+header.firstyear,fmin2,fmax2,cavg/header.ncell);
        }
      }
    }
    fclose(file);
    free(vec);
    if(iserr)
    {
      free(unit);
      unit=NULL;
      continue;
    }
    if(iscsv)
    {
      if(verbose)
        printf("%s,0,%g,%g,%g\n",argv[i],fmin,fmax,favg/header.nyear);
      else
        printf("%s,%g,%g,%g\n",argv[i],fmin,fmax,favg/header.nyear);
    }
    else
    {
      if(argc-iarg>1)
        printf("%s: ",argv[i]);
      if(unit!=NULL && strlen(unit))
        printf("min=%g %s, max=%g %s, avg=%g %s\n",fmin,unit,fmax,unit,favg/header.nyear,unit);
      else
        printf("min=%g, max=%g, avg=%g\n",fmin,fmax,favg/header.nyear);
    }
    free(unit);
    unit=NULL;
  }
  return EXIT_SUCCESS;
} /* of 'main' */
